tools/system_libs.py

```diff
@@ -286,325 +605,391 @@ def create_libc(libname):
+  name = 'libc++'
+  symbols = read_symbols(shared.path_from_root('system', 'lib', 'libcxx', 'symbols'))
+  depends = ['libc++abi']
+
+  includes = ['system', 'lib', 'libcxxabi', 'include']
+  cflags = ['-std=c++11', '-DLIBCXX_BUILDING_LIBCXXABI=1', '-D_LIBCPP_BUILDING_LIBRARY', '-Oz',
+            '-D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS']
+
+  src_dir = ['system', 'lib', 'libcxx']
+  src_files = [
+    'algorithm.cpp',
+    'any.cpp',
+    'bind.cpp',
+    'chrono.cpp',
+    'condition_variable.cpp',
+    'debug.cpp',
+    'exception.cpp',
+    'future.cpp',
+    'functional.cpp',
+    'hash.cpp',
+    'ios.cpp',
+    'iostream.cpp',
+    'locale.cpp',
+    'memory.cpp',
+    'mutex.cpp',
+    'new.cpp',
+    'optional.cpp',
+    'random.cpp',
+    'regex.cpp',
+    'shared_mutex.cpp',
+    'stdexcept.cpp',
+    'string.cpp',
+    'strstream.cpp',
+    'system_error.cpp',
+    'thread.cpp',
+    'typeinfo.cpp',
+    'utility.cpp',
+    'valarray.cpp',
+    'variant.cpp',
+    'vector.cpp',
+    os.path.join('experimental', 'memory_resource.cpp'),
+    os.path.join('experimental', 'filesystem', 'directory_iterator.cpp'),
+    os.path.join('experimental', 'filesystem', 'path.cpp'),
+    os.path.join('experimental', 'filesystem', 'operations.cpp')
+  ]
+
+
+class libmalloc(MTLibrary):
+  name = 'libmalloc'
+
+  cflags = ['-O2', '-fno-builtin']
+
+  def __init__(self, **kwargs):
+    self.malloc = kwargs.pop('malloc')
+    if self.malloc not in ('dlmalloc', 'emmalloc', 'none'):
       raise Exception('malloc must be one of "emmalloc", "dlmalloc" or "none", see settings.js')
 
-    # only dlmalloc supports most modes
-    def require_dlmalloc(what):
-      if base != 'dlmalloc':
-        shared.exit_with_error('only dlmalloc is possible when using %s' % what)
-
-    extra = ''
-    if shared.Settings.DEBUG_LEVEL >= 3:
-      extra += '_debug'
-    if not shared.Settings.SUPPORT_ERRNO:
-      # emmalloc does not use errno anyhow
-      if base != 'emmalloc':
-        extra += '_noerrno'
-    if shared.Settings.USE_PTHREADS:
-      extra += '_threadsafe'
-      require_dlmalloc('pthreads')
-    if shared.Settings.EMSCRIPTEN_TRACING:
-      extra += '_tracing'
-      require_dlmalloc('tracing')
-    if base == 'dlmalloc':
-      source = 'dlmalloc.c'
-    elif base == 'emmalloc':
-      source = 'emmalloc.cpp'
-    return (source, 'lib' + base + extra)
-
-  def malloc_source():
-    return malloc_decision()[0]
-
-  def malloc_name():
-    return malloc_decision()[1]
-
-  def create_malloc(out_name):
-    if not malloc_source():
-      return
-    o = in_temp(out_name)
-    cflags = ['-O2', '-fno-builtin']
-    if shared.Settings.USE_PTHREADS:
-      cflags += ['-s', 'USE_PTHREADS=1']
-    if shared.Settings.EMSCRIPTEN_TRACING:
-      cflags += ['--tracing']
-    if shared.Settings.DEBUG_LEVEL >= 3:
+    self.is_debug = kwargs.pop('is_debug')
+    self.use_errno = kwargs.pop('use_errno')
+    self.is_tracing = kwargs.pop('is_tracing')
```

Comment:
What would happen if some abstract library wanted to use the same kwargs? Should these have a `malloc_` prefix or something like that?

Reply:
That would only be a problem if `libmalloc` inherits from that abstract library. If that were the case, we probably wouldn't need this line.

If we really want to, we _could_ use the `__` name mangling, but I don't feel it's necessary.